### PR TITLE
Refactor: state-aware ViewModel and app flow 

### LINKED ISSUE #23: Connections and permissions state machine
Currently, we have a few bugs:
 - The scan button does not always transition to "enable permission" when an adb connection is lost
 - Auto reconnect on app or devie restart is not well handled
 - https://github.com/osservatorionessuno/bugbane/issues/11

We have a few settings and requirements that should be checked.
 - Is Wifi connected (pre requisite for everything)
 - Is notification permission granted
 - Are Developer Options enabled
 - Is `_adb-tls-connect._tcp` registered (signal Wireless ADB is on)
 - Is `_adb-tls-pairing._tcp` registered (signal Wireless ADB is in pairing mode, temporarily replaces 'connect')
 - Is adb connected

Now, in the scanning screen, we should have a callback to know whether `_adb-tls-connect._tcp` is registered or unregistered.
 - If registered:
     - If an adb connection as already established, probe it (show a spinner)
     - If an adb connection was not already established, try to auto connect (show a spinner)
  - If unregistered:
    - If an adb connection was previously established, drop it
   

The slideshow should run on startup `_adb-tls-connect._tcp` is unregistered and there are no pre-existing acquisitions.

In the slideshow, we support skipping individual slides: for instance is wifi is already enabled, there is no need to show the wifi slide. If notification permission is already granted, there is no need to show the slide to request permissions.

The pairing slide is a bit more tricky: if the user arleady paired in the past, they may just need to enable Wifi Debugging and not go through the pairing process again (although doing it would work). Currently the flow assumes that the user guided to enable Wifi debugging will also always do the pairing. It is a relatively niche case but we should eventually fix it.

### PR DESCRIPTION
**Pull request by rocodes**
(WIP)  Still needs work and bugfixes but posting for feedback since the concept/structure is there (cc @TheZ3ro).  
It's....kind of a lot?

- Define AppState and refactor Slideshow to avoid individual pages with individual logic, instead having a single  SlideshowPage that displays the right content for a given state. 
- State logic (requisites, forward/back) are all controlled in the (new) ConfigurationViewModel.  
- In addition to the checks like isDeveloperOptionsEnabled, is WifiEnabled, the app state can also observe other components and be responsive to their changes: right now, one for adb connectivity status and one rough one for wifi connectivity status -- so for example disabling wifi mid-flow doesn't require a check on every page, it will emit the NeedsWifi state and the app will show the slideshow starting at wifi (and skipping any states it doesn't need again).
- AdbViewModel is no longer a viewmodel. AdbManager.java was ported to Kotlin, and followup work can condense/combine some of the adb connector classes as @TheZ3ro suggested, but already this was getting too big so I mostly left them alone, except for replacing individual mutable data objects (isPair, isConnected, etc) with an AdbState stateflow. But there are a ton of adb service bootstrapping classes that can probably be cut down a lot.

Still todo:
- Resolve some bugs around returning from/to different menus, around finishing the slideshow after pairing
- adbState / the Scan Screen logic is incomplete 

Fixes #23 
Refs #11 (will close once finished)
Towards #12

### CODE REVIEW

**Comment by lsd-cat** [comment]
Ci fails is my fault, hopefully improvements will come up in the next few days. This looks great and so much! I think we will benefit a lot from a more solid design here in the long term, and it will probably solve a lot of edge cases we have not even experienced yet! Thanks so much and looks forward to look into more detail and test it.

File: app/src/main/java/org/osservatorionessuno/bugbane/MainActivity.kt [hunk 1]
```diff
@@ -30,64 +34,64 @@ import org.osservatorionessuno.bugbane.components.NavigationTabs
...
+                        Log.d(TAG, "TryAutoConnect")
+                        configViewModel.adbManager.autoConnect()
+                        // Else-if, to avoid conflict with slideshow connection logic
+                    }
+                }
+
+                if (appState.value != AppState.TryAutoConnect) {
+                    MainContent()
+                } else { // Trying autoconnect, which will either lead to success or need repairing
+                    WaitLoadingContent()
```
Comment (rocodes) [inline app/src/main/java/org/osservatorionessuno/bugbane/MainActivity.kt:80]: (note: We can deprecate this WaitLoadingContent() part, which basically blocks the UI until the adb autoconnection has either succeeded or failed. The reason it's here is that I wanted to isolate the changes in this refactor (for testing) before adding additional logic (like disabling buttons until autoconnect is completed)

**Comment by rocodes** [comment]
I noticed there's a regression in the pairing flow so I'm going to put this back in draft mode while I fix it :(

**Comment by rocodes** [comment]
ok @TheZ3ro I think I'm ready for you now :)

File: app/src/main/java/org/osservatorionessuno/bugbane/screens/ScanScreen.kt [hunk 2]
```diff
@@ -361,12 +362,13 @@ fun ScanScreen(
...
-                        text = if (isScanning)
-                            stringResource(R.string.home_scanning_button)
-                        else if (lacksPermissions)
-                            stringResource(R.string.home_permissions_button)
-                        else
-                            stringResource(R.string.home_scan_button),
+                        text = when (appState.value) {
+                            AppState.AdbScanning -> stringResource(R.string.home_scanning_button)
+                            AppState.AdbConnected -> stringResource(R.string.home_scan_button)
+                            AppState.TryAutoConnect, AppState.AdbConnecting -> stringResource(R.string.notification_adb_pairing_working_title)
```
Comment (rocodes) [inline app/src/main/java/org/osservatorionessuno/bugbane/screens/ScanScreen.kt:368]: (as with above, `notification_adb_pairing_working_title` could be improved on here to ~ "trying to connect to adb" but just tried to reuse existing strings and avoid more translation burden until things are more finalized)

**Comment by TheZ3ro** [comment]
Ok wow I've understood the problem here.

Basically, when you disable developer options "USB debugging" and "Wireless debugging" gets disabled.
Then during the Slideshow, we instruct the user to enable "Wireless debugging".
If "USB debugging" is **not enabled**, then pairing is successful but ADB-wireless doesn't seems to be able connect!
Turing it on make everything smoothless! 

image

**Comment by rocodes** [comment]
Thank you both for your feedback :)  

@lsd-cat: Thanks for the video. It looks like it's trying to autoconnect first (which is why the button text says "Pairing with adb" and then switches to "Pair"), then it fails and asks you to re-pair. If you're starting fresh (no prior Bugbane connections stored in Wireless Debugging) it _shouldn't_ happen, but maybe there's a case I didn't catch, I'll try to repro. 

@TheZ3ro: clarification about what you found, are you testing on a physical device by running the app while it's plugged and in usb debug mode, or are you seeing this issue even when you push the apk to the device via usb, then disconnect the phone, then run the app?  I'll try to reproduce it as well, or if you have steps to reproduce the failed connection issue it would be helpful too.

I think I know what the last couple issues are, thanks for the testing and I'll see what I can do.

**Comment by rocodes** [comment]
I think I just realized another potential improvement when I was looking at the adb (usb) state vs the adb_wifi state: we can use contentobservers to know when there are changes. I'm going to experiment with some more changes, I know it's maybe a lot but I think it's worth it to try to get the logic part clean upfront.

File: app/src/main/java/org/osservatorionessuno/bugbane/utils/ConfigurationViewModel.kt [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+    }
+
+
+    internal fun getIntentForAppState(state: AppState): Intent? {
+        return when (state) {
+            AppState.NeedWifi -> Intent(Settings.ACTION_WIFI_SETTINGS)
+            AppState.NeedNotificationConfiguration -> Intent(Settings.ACTION_APP_NOTIFICATION_SETTINGS).apply {
+                putExtra(Settings.EXTRA_APP_PACKAGE, appContext.packageName)
+            }
+            AppState.NeedDeveloperOptions -> Intent(Settings.ACTION_DEVICE_INFO_SETTINGS)
```
[the remark this answers is not in the document]
Comment (rocodes) [reply app/src/main/java/org/osservatorionessuno/bugbane/utils/ConfigurationViewModel.kt]: Yeah, so I basically did what we do for the wireless debugging intent (pass extras and try to highlight the right key in the [right menu fragment](https://cs.android.com/android/platform/superproject/main/+/main:packages/apps/Settings/res/xml/my_device_info.xml)) - it's a bit brittle and would be nice  to confirm that at worst it fails silently if the key is wrong, so we should test on a Samsung or something where the menu is a bit different.

**Comment by rocodes** [comment]
hm. thanks for the detailed steps, I'll look into it more, will take it back into draft mode for a bit, _but_:  because the best ideas clearly always come after one has already put up a PR, I kind of reworked a bunch of this. (still wip / not working yet but I think it will be an improvement:) 

- [make the configuration checks ContentObservable](https://github.com/osservatorionessuno/bugbane/commit/45a4d0935af2b9eaf66c2a9aa94536f847b5efa4) and listen for observavle updates instead of querying for them 
 
- [ simplify the state checking logic a lot](https://github.com/osservatorionessuno/bugbane/commit/80873de9948a4bc9aae89618a9215f056edf3436) 

then the viewmodel actually just emits state, like a viewmodel and it gets rid of all the `checkUpdateState` stuff and just behaves nicely. 

I will clean it all up + push it + hopefully this squashes all the sad weird states problems

File: app/src/main/java/org/osservatorionessuno/bugbane/components/SlideshowPage.kt [hunk 4]
```diff
@@ -1,34 +1,102 @@
...
+        )
+        AppState.NeedWirelessDebuggingAndPair -> return SlideshowPageData(title = stringResource(R.string.slideshow_wireless_and_pair_title),
+            description = stringResource(R.string.slideshow_wireless_and_pair_description),
+            icon = Icons.Filled.Build,
+            buttonText = stringResource(R.string.slideshow_wireless_and_pair_button),
+        )
+        AppState.TryAutoConnect, AppState.AdbConnecting, -> return SlideshowPageData(title = stringResource(R.string.notification_channel_adb_pairing), //todo
+            description = stringResource(R.string.notification_adb_pairing_working_title),
+            icon = Icons.Filled.Build,
+            buttonText = "please wait", //todo
```
[the remark this answers is not in the document]
Comment (lsd-cat) [reply app/src/main/java/org/osservatorionessuno/bugbane/components/SlideshowPage.kt:79]: Thanks! I'd go for "Waiting for ADB pairing" and "In attesa dell'accoppiamento ADB" (@TheZ3ro yes Android uses "accoppiamento" a lot, [see this](https://android.googlesource.com/platform/frameworks/base/+/e92cf6d/packages/SettingsLib/res/values-it/strings.xml)). 

I think I'd keep the button clickable though, for the following scenario: if the user exits the pairing process, or closes the application, than at the next start of the app it is still waiting for pairing, but the user has no way to get redirected again to the correct screen (though the app behavior seemed very solid, keeping the pairing service running or restarting it and allowing the user to complete the flow no matter in which state they interrupted).

[the remark this answers is not in the document]
Comment (rocodes) [reply app/src/main/java/org/osservatorionessuno/bugbane/components/SlideshowPage.kt]: thank you :) added the text in 5f3f356, and also pushed a commit to make sure the pairing service notification is killed if the app is exited via the slideshow.

All that by way of saying, if you can get things into a state where the user is stuck on a screen that doesn't automatically resolve to a connection success or error (and then the pairing screen), that's a state machine bug that I should fix. The connecting / tryautoconnect states are always supposed to automatically resolve to a state where the user can take actions.

**Review by TheZ3ro — APPROVED** [review]
This is much much more reliable and elegant than the old `canSkip` Slideshow, I really like this approach. LGTM

File: app/src/main/java/org/osservatorionessuno/bugbane/MainActivity.kt [hunk 5]
```diff
@@ -30,64 +28,53 @@ import org.osservatorionessuno.bugbane.components.NavigationTabs
...
+                        val startPage = (appState.value.step)
+                        val intent = Intent(this@MainActivity, SlideshowActivity::class.java)
+                            .putExtra("startPage", startPage)
+                        startActivity(intent)
+                    }
                 }
             }
         }
+
+        configViewModel.adbManager.watchCommandOutput().observe(this) { output ->
```
[the remark this answers is not in the document]
Comment (TheZ3ro) [reply app/src/main/java/org/osservatorionessuno/bugbane/MainActivity.kt:73]: sorry I was wrong, that is the listener on the output of the QuickForensics result.
maybe we can just clean that out
